### PR TITLE
Widen the accumulator for integer reduce_mean under XLA

### LINKED ISSUE #125978: tf.reduce_mean on int32/uint32 overflows under jit_compile=True and returns a negative mean of positive inputs
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

2.15.0; tf-nightly 2.22.0-dev20260823

### Custom code

Yes

### OS platform and distribution

macOS 26.0, Apple Silicon (arm64)

### Mobile device

_No response_

### Python version

3.10 (TF 2.15.0); 3.12.10 (tf-nightly)

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

N/A — CPU only

### Current behavior?

`tf.reduce_mean` on an `int32` tensor returns a different—and mathematically wrong—result under `tf.function(jit_compile=True)` than in eager execution when the sum of the reduced elements overflows the element type.

```python
a = tf.constant([[2**31 - 1, 2**31 - 1]], dtype=tf.int32)
tf.reduce_mean(a, axis=1)
# eager                    -> [2147483647]
# tf.function(jit_compile) -> [-1]
```

One operator and two elements are enough. The JIT result is consistent with accumulating in the element dtype, wrapping, and then dividing: it returns a **negative mean for strictly positive inputs**.

### Expected behavior

The compiled result should agree with eager execution and return `[2147483647]`.

### Scope

The full matrix below was measured on TensorFlow 2.15.0, CPU. The minimal `int32` and `uint32` cases, as well as the duplication example below, were also confirmed on `tf-nightly 2.22.0-dev20260823` (Git version `v1.12.1-143961-g64fbb70af66`).

| condition | result |
|---|---|
| `int32`, `uint32` | **wrong** |
| `int8`, `int16`, `int64`, `uint8`, `uint16`, `uint64` | agrees with eager |
| `reduce_mean` | **wrong** |
| `reduce_sum`, `reduce_max`, `reduce_prod` | agree with eager |
| `float32`, `float64` | agree with eager |
| `jit_compile=True` | **wrong** |
| autoclustering (`tf.config.optimizer.set_jit(True)`) | agrees with eager |
| values that do not overflow | agrees with eager |
| `axis=0`, `axis=1`, whole-tensor reduce | all wrong |
| repeated runs | deterministic, 5/5 |

For the overflowing `int64` control, both paths return `-1`; it therefore does not produce an eager-versus-compiled discrepancy.

A further consequence is that `reduce_mean` under JIT is not invariant under data duplication. These expressions have the same mathematical mean, but only the second has an overflowing internal sum:

```python
half = (2**31 - 1) // 2
v = tf.constant([[half, half + 1]], dtype=tf.int32)
tf.reduce_mean(v, axis=1)                          # jit and eager: 1073741823
tf.reduce_mean(tf.concat([v, v], axis=1), axis=1) # jit: 0, eager: 1073741823
```

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf


def mean1(t):
    return tf.reduce_mean(t, axis=1)


def check(label, fn, *args):
    eager = fn(*args).numpy()
    compiled = tf.function(
        fn, jit_compile=True, autograph=False
    )(*args).numpy()
    print(
        f"{label:40s} eager={eager}  jit={compiled}  "
        f"{'ok' if np.array_equal(eager, compiled) else 'MISMATCH'}"
    )


print("tensorflow", tf.__version__)

M = 2**31 - 1
a = tf.constant(np.array([[M, M]], dtype=np.int32))

check("int32 [[MAX, MAX]] reduce_mean", mean1, a)
check(
    "int32 [[10, 20]] reduce_mean",
    mean1,
    tf.constant(np.array([[10, 20]], dtype=np.int32)),
)
check(
    "int32 reduce_sum (control)",
    lambda t: tf.reduce_sum(t, axis=1),
    a,
)

for dt in (
    "int8", "int16", "int32", "int64",
    "uint8", "uint16", "uint32", "uint64",
):
    i = np.iinfo(dt)
    check(
        f"{dt} reduce_mean",
        mean1,
        tf.constant(np.array([[i.max, i.max]], dtype=dt)),
    )
```

### Relevant log output

```shell
TensorFlow Nightly:

tensorflow 2.22.0-dev20260823
Git version v1.12.1-143961-g64fbb70af66
int32:  eager=[2147483647]  jit=[-1]          MISMATCH
uint32: eager=[4294967295]  jit=[2147483647]  MISMATCH
dup-base:   eager=[1073741823] jit=[1073741823] ok
dup-concat: eager=[1073741823] jit=[0]          MISMATCH

TensorFlow 2.15.0:

int32 [[MAX, MAX]] reduce_mean           eager=[2147483647]  jit=[-1]  MISMATCH
int32 [[10, 20]] reduce_mean             eager=[15]          jit=[15] ok
int32 reduce_sum (control)               eager=[-2]          jit=[-2] ok
int8 reduce_mean                         eager=[127]         jit=[127] ok
int16 reduce_mean                        eager=[32767]       jit=[32767] ok
int32 reduce_mean                        eager=[2147483647]  jit=[-1] MISMATCH
int64 reduce_mean                        eager=[-1]          jit=[-1] ok
uint8 reduce_mean                        eager=[255]         jit=[255] ok
uint16 reduce_mean                       eager=[65535]       jit=[65535] ok
uint32 reduce_mean                       eager=[4294967295]  jit=[2147483647] MISMATCH
uint64 reduce_mean                       eager=[9223372036854775807] jit=[9223372036854775807] ok
```

### PR DESCRIPTION
**Pull request by vishwakt**
Fixes #125978

`tf.reduce_mean` on `int32` and `uint32` disagrees with eager execution under `tf.function(jit_compile=True)` when the sum of the reduced elements overflows the element type, returning a negative mean for strictly positive inputs:

```python
a = tf.constant([[2**31 - 1, 2**31 - 1]], dtype=tf.int32)
tf.reduce_mean(a, axis=1)
# eager                    -> [2147483647]
# tf.function(jit_compile) -> [-1]
```

### Cause

`MeanOp` in `tensorflow/compiler/tf2xla/kernels/reduction_ops.cc` accumulates in `XlaHelpers::SumAccumulationType(input_type)` and then divides. That helper widens only the narrow integer types:

```cpp
// Upcast small integer types to 32 bit to avoid overflow.
if (dtype == DT_INT8 || dtype == DT_INT16) return DT_INT32;
if (dtype == DT_UINT8 || dtype == DT_UINT16) return DT_UINT32;
return dtype;
```

`int32` and `uint32` fall through unwidened, so the sum wraps before the division is applied. This accounts for the reported behavior exactly: `int8` and `int16` widen to `int32` and agree with eager, `uint8` and `uint16` widen to `uint32` and agree, `int64` and `uint64` have nothing wider and wrap in eager too so they also agree, and `int32` and `uint32` are the only two that diverge.

It is also why `reduce_sum` agrees while `reduce_mean` does not. Sum returns the wrapped sum either way, since the result is converted back to the input type at the end. Only Mean divides, which turns a wrapped accumulator into a quotient that is wrong in both sign and magnitude.

### Fix

`Mean` now selects its own accumulation type, widening the 32 bit integer types and deferring to `SumAccumulationType` for everything else:

```cpp
DataType MeanAccumulationType(const DataType& dtype) {
  if (dtype == DT_INT32) return DT_INT64;
  if (dtype == DT_UINT32) return DT_UINT64;
  return XlaHelpers::SumAccumulationType(dtype);
}
```

`SumAccumulationType` itself is deliberately left alone. It feeds roughly sixteen other call sites, including `Sum`, `Prod`, `L2Loss`, softmax, LRN, batch norm, pooling and the scan ops, and none of those need a wider accumulator to match eager. Confining the change to `Mean` fixes the divergence without altering the accumulation width of every other reduction.

`XlaReductionOp` already converts the input to the reduction type before reducing, so no other change is needed for the wider accumulator to take effect.

`int64` and `uint64` continue to wrap, which matches the eager kernel: for those types the reporter's own control shows both paths returning `-1`, so there is no divergence to fix and nothing wider to accumulate in.

### Tests

`testReduceMeanIntegerOverflow` in `tensorflow/compiler/tests/reduce_ops_test.py` reduces two copies of the largest representable value for `int32` and `uint32` and checks the result equals that value, guarded on the dtype being supported by the backend. It follows the existing class-level index dtype parameterization.

Verified against the released 2.21.0 wheel that the assertion the test makes fails on the current kernels:

```
int32:  expected [2147483647]  got [-1]
uint32: expected [4294967295]  got [2147483647]
```

### Notes for reviewers

- I could not build TensorFlow locally, so the C++ change has not been compiled and the new test has not been executed. The reproducer, the dtype matrix and the gating evidence above were verified against the released 2.21.0 wheel on macOS arm64 CPU.
- The tradeoff is that integer `Mean` now accumulates in 64 bit for 32 bit inputs. The existing comment in `SumAccumulationType` indicates the widening is there to avoid overflow in the first place, so this extends the stated intent rather than introducing a new policy, but I am happy to reconsider if the cost is a concern for large integer reductions.
- Worth noting for anyone looking at the MLIR path: the reduction lowering there uses a different helper, `GetAccumulationType` in `tensorflow/compiler/mlir/tf2xla/transforms/legalize_tf.cc`, which widens no integer types at all. The observed results identify the kernel above as the one that is live for this case, since the MLIR path would have returned -1 for `int8` rather than the 127 that is actually produced.

### CODE REVIEW

File: tensorflow/compiler/tf2xla/kernels/reduction_ops.cc [hunk 1]
```diff
@@ -123,11 +123,25 @@ class MaxOp : public XlaReductionOp {
...
+// Mean divides the accumulated sum by the number of reduced elements, so an
+// accumulator that wraps does not merely produce a wrapped sum the way Sum
+// does, it produces a quotient that is wrong in both sign and magnitude. That
+// is why Mean needs a wider accumulator than the other reductions:
+// `SumAccumulationType` already widens the 8 and 16 bit integer types to avoid
+// overflow, but leaves the 32 bit ones alone, so reducing int32 or uint32 gave
+// a negative mean for strictly positive inputs and disagreed with the eager
+// kernel. The 64 bit types have nothing wider to accumulate in and keep
+// wrapping, which is what the eager kernel does for them as well.
+DataType MeanAccumulationType(const DataType& dtype) {
```
[the remark this answers is not in the document]
Comment (vishwakt) [reply tensorflow/compiler/tf2xla/kernels/reduction_ops.cc:135]: This function is already inside the anonymous namespace in this file, which opens at line 37 and closes at line 221, so it has internal linkage and is not visible at global scope. There is no linker collision to avoid here.

The unnamed namespace and `static` are alternatives for internal linkage rather than things to combine, and the surrounding code agrees: free functions defined inside anonymous namespaces in tensorflow/compiler/tf2xla/kernels are plain in about 127 cases against 11 that add `static`, so leaving it unqualified matches the convention in this directory. Leaving it as is.

**Review by dmiltr3 — APPROVED** [review]
